### ISSUE TITLE
Implement better Errors

### ISSUE THREAD

**Issue by davesnx** [davesnx/styled-ppx#429]
## Current errors

The current errors are insuficient to learn how to resolve the issue when the parser is complaining, and lack a bit of context, they are classified by:

### Lexing/Parsing errors

The current parser errors regarding variables and missing features is very bad.

![](https://user-images.githubusercontent.com/3474647/91650346-63bd8400-ea4c-11ea-865b-6abf182b2636.png)

![](https://user-images.githubusercontent.com/3763599/76766752-c2da5580-6798-11ea-86dc-85b932823be7.png)

### Properties raises

- Unsupported feature

### ppx raise

- API error usage

## Improving Parser errors

Maybe we can improve ParseError and display the error with the expected next values?

### Type-check errors

Render error messages with the value ppx payload

![E9yxbcbXMAE7Nlc.jpeg](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/352cef03-691e-4a91-a705-dfc9912503a5/E9yxbcbXMAE7Nlc.jpeg)

### Wrong value

```jsx
display: blocki;
         ^^^^^^

Not a valid value for `display`.
display defintion expects one of `'block' | 'contents' | 'flex' | 'flow' | 'flow-root' | 'grid' | 'inline' | 'inline-block' | 'inline-flex' | 'inline-grid' | 'inline-list-item' | 'inline-table' | 'list-item' | 'none' | 'ruby' | 'ruby-base' | 'ruby-base-container' | 'ruby-text' | 'ruby-text-container' | 'run-in' | 'table' | 'table-caption' | 'table-cell' | 'table-column' | 'table-column-group' | 'table-footer-group' | 'table-header-group' | 'table-row' | 'table-row-group' | '-ms-flexbox' | '-ms-inline-flexbox' | '-ms-grid' | '-ms-inline-grid' | '-webkit-flex' | '-webkit-inline-flex' | '-webkit-box' | '-webkit-inline-box' | '-moz-inline-stack' | '-moz-box' | '-moz-inline-box'`
(optional) Maybe you mean `block`?
```

### Wrong value format

Detect when partial matching happens and render the closest as an error message

```jsx
border: 1px solid;
^^^^^^^^^^^^^^^^^

Missing value for `border`.
border defintion is `<line-width> || <line-style> || <color>`
Missing `<color>`

// ----
// TODO: More examples
```

### Wrong type

```jsx
border-width: 1%;
			  ^^

Wrong value type for `border-width`
border-width definition is `[ <line-width> ]{1,4}`. (For example: `2px`)
```

### Missing property

When a not valid property is passed to our parser (we should differentiate when the property is a valid CSS but we don't support vs a invalid property)

```jsx
border-width-left: 1%;
^^^^^^^^^^^^^^^^^

Invalid property `border-width-left`.
(optional) Maybe you mean `border-left-width`?
```

### Unsupported interpolation in property

```jsx
resize: $(lola);
        ^^^^^^^

`resize` property doesn't support interpolation `$(lola)`. 
```

## Resources

Check for error reporting from mehir and driver: https://github.com/Innf107/polaris

**Comment by davesnx** [comment]
```
File "xxx/PasswordStrength_Css.ml", line 16, characters 8-9:
Error: Parse error while reading token '['
```

Actual issue is on line 170, incorrect selector:
```
  hover [
    opacity: 0.7;
  ];
```

![CleanShot 2024-03-08 at 13 32 06@2x](https://github.com/davesnx/styled-ppx/assets/3763599/1f4c833e-c2da-436d-b13f-6bb52fe1dcdf)

**Comment by davesnx** [comment]
```
File "frontend/packages/ahkit/src/bindings/PasswordStrength_Css.ml", line 79, characters 25-26:
Error: Parse error while reading token ';'
``` 

```
mask-image: $("url(" ^ eyeCrossedIcon ^ ")");
```

![CleanShot 2024-03-08 at 13 23 22@2x](https://github.com/davesnx/styled-ppx/assets/3763599/799f69d1-d073-4a08-9f80-f6fb603773ea)
